### PR TITLE
Add profiler support in llm foundry

### PR DESCRIPTION
**Pull request by j316chuck**
# Description
Add profiler support for llm foundry

## Along for the ride:

Adding a yaml to support training mpt models in CPU mode. Ths is useful so you don't have to spin up an interactive session, wait for an interactive session download, set up interactive session, get/download your data, and let's you test quickly on a small model for cpu only features. Only con is no gpu 😛 

# Tests
`composer train/train.py \
  train/yamls/pretrain/mpt-small-cpu.yaml \
  data_local=my-copy-c4 \
  train_loader.dataset.split=train_small \
  eval_loader.dataset.split=val_small \
  max_duration=10ba \
  eval_interval=0 \
  save_folder=mpt-125m`
  
Produces the chrome traces: `composer_traces/ep0-ba6-rank0.json.` Example:  
```
 {"ph": "X", "cat": "python_function", "name": "queue.py(213): _put", "pid": 0, "tid": 1981795, "ts": 1697505580241206, "dur": 0, "args": {"Ev Idx": 2178228, "Python id": 1145107, "Python parent id": 1145104}},
{"ph": "X", "cat": "python_function", "name": "<built-in method append of collections.deque object at 0x2c5582890>", "pid": 0, "tid": 1981795, "ts": 1697505580241206, "dur": 0, "args": {"Ev Idx": 2178229, "Python id": 1145108, "Python parent id": 1145107}},
{"ph": "X", "cat": "python_function", "name": "threading.py(359): notify", "pid": 0, "tid": 1981795, "ts": 1697505580241207, "dur": 0, "args": {"Ev Idx": 2178230, "Python id": 1145109, "Python parent id": 1145104}},
{"ph": "X", "cat": "python_function", "name": "threading.py(279): _is_owned", "pid": 0, "tid": 1981795, "ts": 1697505580241207, "dur": 0, "args": {"Ev Idx": 2178231, "Python id": 1145110, "Python parent id": 1145109}},

```
 
Produces the pytorch traces: `torch_traces/rank0.6.pt.trace.json`. Example:
```
  {
    "ph": "X", "cat": "python_function", "name": "/Users/chuck.tang/composer/composer/utils/string_enum.py(69): __eq__", "pid": 54200, "tid": 1981795,
    "ts": 1697505581229305, "dur": 0,
    "args": {
      "Ev Idx": 2233416, "Python id": 1200295, "Python parent id": 1200286
    }
  },
```
Useful for profiling memory and time usage 
 
https://github.com/mosaicml/llm-foundry/assets/13524881/7195a528-afa4-4199-a84f-d48cb2630b7b

Perfetto View:
Screenshot 2023-10-16 at 9 35 07 PM

S3: 
```
loggers:
  s3: {bucket_uri: s3://mosaicml-internal-checkpoints-shared/ }
```

`aws s3 cp --recursive s3://mosaicml-internal-checkpoints-shared/chuck/mpt_causal_lm_cpu/traces/`

Full training run: 
`mpt-7b-gpu-8-chinchilla-light-profile-ynjNZ2, mpt-7b-gpu-8-chinchilla-full-profile-uJSCOF, mpt-7b-gpu-8-chinchilla-none-profile-Cwm3GA`

### CODE REVIEW

**Review by mvpatel2000 — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -220,6 +220,16 @@ composer train/train.py \
   eval_interval=0 \
   save_folder=mpt-125m
 
+# To iterate on training on CPU run command (dev only):
```
Comment (dakinggg) [inline README.md:223]: Can we not add this to the readme? Would prefer to keep on simple path through the quickstart?

File: scripts/train/yamls/pretrain/mpt-125m-cpu.yaml [hunk 2]
```diff
@@ -0,0 +1,154 @@
+data_local: ./my-copy-c4
```
[the remark this answers is not in the document]
Comment (j316chuck) [reply scripts/train/yamls/pretrain/mpt-125m-cpu.yaml:1]: I can remove the profiler parts but i think having a cpu yaml for people to dev on would be really nice

File: scripts/train/train.py [hunk 3]
```diff
@@ -454,6 +456,23 @@ def main(cfg: DictConfig) -> Trainer:
         for name, logger_cfg in logger_configs.items()
     ] if logger_configs else None
 
+    # Profiling
+    profiler: Optional[Profiler] = None
```
[the remark this answers is not in the document]
Comment (j316chuck) [reply scripts/train/train.py:460]: fixed

File: README.md (hunk 1 above)
Comment (j316chuck) [reply README.md:223]: removed.

File: llmfoundry/utils/builders.py [hunk 4]
```diff
@@ -133,6 +134,8 @@ def build_logger(name: str, kwargs: Dict[str, Any]) -> LoggerDestination:
         return MLFlowLogger(**kwargs)
     elif name == 'inmemory':
         return InMemoryLogger(**kwargs)
+    elif name == 's3':
```
Comment (dakinggg) [inline llmfoundry/utils/builders.py:137]: what is going on here?

Comment (j316chuck) [reply llmfoundry/utils/builders.py:137]: I need to add an s3 logger: 
```
loggers:
  s3: {bucket_uri: s3://mosaicml-internal-checkpoints-shared/ }

```

Comment (j316chuck) [reply llmfoundry/utils/builders.py:137]: https://github.com/mosaicml/composer/blob/e32ac0f460d8092681bae622fdcf21197b586d95/composer/profiler/torch_profiler.py#L129-L131

Comment (dakinggg) [reply llmfoundry/utils/builders.py:137]: hmm this does not match the way we support things these days which is that if you pass in a remote path, we automatically whatever objects are needed for you. also the way you've implemented it, it only works for s3, which is not ideal. i'd suggest refactoring composer to check if the trace path is remote and if so create the RUD

Comment (j316chuck) [reply llmfoundry/utils/builders.py:137]: removed

**Pull request merged**